compat/win32/fscache.c

```diff
@@ -145,20 +147,44 @@ static void fsentry_release(struct fsentry *fse)
 	InterlockedDecrement(&(fse->refcnt));
 }
 
+static int xwcstoutfn(char *utf, int utflen, const wchar_t *wcs, int wcslen)
+{
+	if (!wcs || !utf || utflen < 1) {
+		errno = EINVAL;
+		return -1;
+	}
+	utflen = WideCharToMultiByte(CP_UTF8, 0, wcs, wcslen, utf, utflen, NULL, NULL);
+	if (utflen)
+		return utflen;
+	errno = ERANGE;
+	return -1;
+}
+
 /*
- * Allocate and initialize an fsentry from a WIN32_FIND_DATA structure.
+ * Allocate and initialize an fsentry from a FILE_FULL_DIR_INFORMATION structure.
  */
 static struct fsentry *fseentry_create_entry(struct fscache *cache, struct fsentry *list,
-		const WIN32_FIND_DATAW *fdata)
+		PFILE_FULL_DIR_INFORMATION fdata)
 {
 	char buf[MAX_PATH * 3];
 	int len;
 	struct fsentry *fse;
-	len = xwcstoutf(buf, fdata->cFileName, ARRAY_SIZE(buf));
+
+	len = xwcstoutfn(buf, ARRAY_SIZE(buf), fdata->FileName, fdata->FileNameLength / sizeof(wchar_t));
 
 	fse = fsentry_alloc(cache, list, buf, len);
 
-	if (fdata->dwReserved0 == IO_REPARSE_TAG_SYMLINK &&
+	/*
+	 * On certain Windows versions, host directories mapped into
+	 * Windows Containers ("Volumes", see https://docs.docker.com/storage/volumes/)
+	 * look like symbolic links, but their targets are paths that
+	 * are valid only in kernel mode.
+	 *
+	 * Let's work around this by detecting that situation and
+	 * telling Git that these are *not* symbolic links.
+	 */
+	if (fdata->FileAttributes & FILE_ATTRIBUTE_REPARSE_POINT &&
+	    fdata->EaSize == IO_REPARSE_TAG_SYMLINK &&
```

Comment:
Very nice. I should probably squash that comment into the *other* commit, though, namely the one adding the code: 05e20e1d72c0 (mingw: Windows Docker volumes are *not* symbolic links, 2018-04-23)